### PR TITLE
buy-01 frontend 

### PR DESCRIPTION
**Pull request by adrian-lui**
- added home, product listing, seller management pages
- added add/update products modals
- added get, post, put, delete product api request 
- added media service for uploading and validating images
- added user registration seller role option

broken stuff:
- POST /api/media should be permitAll to allow user to upload avatar on registration?
- GET /api/user/{id} is returning 401, now we cannot fetch product owner's name (cannot get user by userId)
- Backend missing product quantity handler
- Backend missing imageId in GET /api/products struct
- I reverted my backend changes to "SELLER" role, so the backend in this branch doesn't handle any seller logic. But frontend has already implemented the seller role logic

### CODE REVIEW

**Comment by maximihajlov** [comment]
Hi @adrian-lui, some of the issues should be fixed in #33. Could you please review #31 and #32, so we can test it all together?

**Comment by adrian-lui** [comment]
> Hi @adrian-lui, some of the issues should be fixed in #33. Could you please review #31 and #32, so we can test it all together?

Yes, reviewed. Can you merge some changes already so main branch won't be so behind?

**Comment by maximihajlov** [comment]
I've merged. Now there's only #33, which is not merged, but it's rebased onto main. 

Could you please check if new fields are in the desired format?

**Comment by adrian-lui** [comment]
> I've merged. Now there's only #33, which is not merged, but it's rebased onto main.
> 
> Could you please check if new fields are in the desired format?

For me the format is fine. I can adapt to it with some frontend changes. but can you add Quantity field to product to #33 too?

**Review by maximihajlov — CHANGES_REQUESTED** [review]
We found several issues in the current frontend implementation during our meeting. 

1. CRUD operations for users are not implemented on the frontend

> Are CRUD operations for Users and Products correctly implemented, and does each user role have the appropriate access levels?

It's required for users to be able to change their information. It was recently implemented on backend and now can be added to the frontend.

2. Error handling is not correct

> Attempt actions that might cause errors, such as registering with an existing email, uploading invalid media formats, or exceeding the media size limit.
>
> Does the application gracefully handle these errors, providing useful feedback to the user?

The UI does not show detailed error messages sometimes. It also resets the form giving the user no chance to fix an input error and resubmit the form again

https://github.com/merpw/lets-play/assets/43277901/ca155477-7411-4273-ae2f-35bb77bfd9ea



3. The UX is not intuitive enough, the UI also needs improvement
  
> Are all the frontend pages functioning as expected, and is the user experience intuitive?

https://github.com/merpw/lets-play/assets/43277901/8dae05a6-5546-49f5-a0fa-e046a7870760

File: frontend/src/app/pages/edit-profile-page/edit-profile-page.component.ts [hunk 1]
```diff
@@ -0,0 +1,71 @@
...
+  }
+
+  showError(errorMessage: string) {
+    this.result = {
+      type: 'error',
+      message: errorMessage,
+    };
+  }
+
+  updateProfile(userDetails: any) {
```
Comment (maximihajlov) [inline frontend/src/app/pages/edit-profile-page/edit-profile-page.component.ts:32]: It's bad idea to use any for methods that you control. We loose type safety.

Comment (maximihajlov) [reply frontend/src/app/pages/edit-profile-page/edit-profile-page.component.ts:32]: It's aslo recommended to use eslint and resolve all warnings/errors that it shows

https://github.com/angular-eslint/angular-eslint

File: frontend/src/app/pages/sign-up-page/signup-form/signup-form.component.ts [hunk 2]
```diff
@@ -1,26 +1,86 @@
...
+    private router: Router,
+    public authService: AuthService
+  ) {}
+
+  ngOnInit(): void {
+    console.log(this.profile);
+    if (this.profile) {
+      this.form.patchValue(this.profile);
+      this.imageUploaded = this.profile.image;
+      this.form.removeControl('password');
```
Comment (maximihajlov) [inline frontend/src/app/pages/sign-up-page/signup-form/signup-form.component.ts:59]: maybe also hide role

**Review by maximihajlov — CHANGES_REQUESTED** [review]

**Review by maximihajlov — APPROVED** [review]
LGTM. Not perfect, but it works 🚀

**Pull request merged**